_posts/2018-09-23-Inter-service-communication.md

```diff
@@ -0,0 +1,208 @@
+
+# Building an advanced custom Feign client
+When we need more advanced tooling inside our Feign client, there is the Feign configuration class where we can build and customize our Feign client. 
+To begin, we just create a new class annotated with `@Configuration`.
+
+{% highlight java %}
+@Configuration
+class SapClientFeignConfig { }
+{% endhighlight %}
+
+## The builder
+Feign provides us with a builder like pattern to build up our client.
+When we want to customize, we just add our own custom classes to the builder. 
+First off, lets create a bean of our client and return a Feign builder.
+It's important to to let the builder know which client he has to target for communication. 
+The second parameter is most likely the base url where all the requests begin. 
+
+{% highlight java %}
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Interceptor
+If you need some basic authorization, custom headers or some extra information in every request of the client, we can use interceptors. 
+This becomes very useful in situations where every requests needs this extra information.
+To add an interceptor, we just add an extra method that returns the Feign interceptor. 
+
+{% highlight java %}
+ private RequestInterceptor requestInterceptor() {
+        return requestTemplate -> {
+            requestTemplate.header("user", username);
+            requestTemplate.header("password", password);
+            requestTemplate.header("Accept", ContentType.APPLICATION_JSON.getMimeType());
+        };
+    }
+{% endhighlight %}
+
+To enable the customization, we add the interceptor to the builder. 
+
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .requestInterceptor(requestInterceptor())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Client
+By default, the default HTTP client of Feign uses HttpUrlConnection to execute its HTTP requests.
+You can configure another client (ApacheHttpClient, OkHttpClient, ...) as follows:
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .client(new ApacheHttpClient())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+#### OkHttpClient
+OkHttp is an HTTP client that’s efficient by default:
+
+* HTTP/2 support allows all requests to the same host to share a socket.
+* Connection pooling reduces request latency (if HTTP/2 isn’t available).
+* Transparent GZIP shrinks download sizes.
+* Response caching avoids the network completely for repeat requests.
+
+#### Mutual SSL with ApacheHttpClient
+The advantage of using ApacheHttpClient over the default client is that ApacheHttpClient sends more headers with the request, eg. Content-Length, which some servers expect.
```

Comment:
I'd personally use more inline code styling for the class names or properties:
`ApacheHttpClient `
`Content-Length`